aws/aws-toolkit-vscode
# Credentials Management

## Description

This change introduces a more managed approach to implementing Credentials Providers for the toolkit. The goal is to make it easier to onboard support for new credentials types moving forward. This design is very similar to the approach taken in the AWS Toolkit for JetBrains.

The design document is in #889

Integrating this new system brought about some functional changes and bug fixes.

New Behavior:
* added logging for Shared Credentials profiles detection and validation
* credentials are now shown by "credential provider id" instead of profile name (example: `profile:default`)
* whenever users run the "Connect to AWS" command, the Toolkit's list of available credentials providers is refreshed (eg: pulling the latest updates from the shared credentials files)
* Credentials are cached and re-used for the duration of a toolkit session. When logging in, if a Shared Credentials Profile has been modified since it was cached, the updated version of the profile will be used instead of the cached version now.
* updated the "Invalid Credentials" notification, and added a button guiding users to the logs
![image](https://user-images.githubusercontent.com/39839589/72020306-5edb6400-3220-11ea-96ec-748204eca4b5.png)

Out of scope:
* a follow-up PR will normalize all credentials related code locations
* a follow-up PR will normalize all user facing text and all code names related to credential vs profile. "profile" terminology will be removed except where explicitly referencing profiles. Previously, the codebase wrongly used these terms rather interchangeably.


## Testing

* additional tests were added to the codebase
* with logging on verbose level:
  * bring up the 'connect to aws' credentials list successive times. See that the logs show that the shared credentials files are only loaded once
  * bring up the 'connect to aws' credentials, then modify the shared credentials files. Bring up the credentials list again. See that the logs show that the shared credentials files were loaded twice
  * set up profiles that make a cyclical reference, and that refer to a nonexistent profile. Bring up the 'connect to aws' credentials list, and see that these profiles are not in the list (and that their validation issues are logged)
  * log in with profile X. Edit the credentials file to alter profile X. log in again with profile X in the same toolkit session. See that the logs show the profile was re-loaded and the new version of X was used to make a connection.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
